managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1850,9 +1854,172 @@ public void offloadFailed(ManagedLedgerException e, Object ctx) {
+                                (ignore2, exception) -> {
+                                    if (exception != null) {
+                                        callback.offloadFailed(new ManagedLedgerException(exception), ctx);
+                                    } else if (offloadException != null && errorsToReport > 0) {
+                                        callback.offloadFailed(new ManagedLedgerException(offloadException), ctx);
+                                    } else {
+                                        callback.offloadComplete(firstUnoffloaded, ctx);
+                                    }
+                                });
+                    } else {
+                        // All failed, so throw up the error
+                        callback.offloadFailed(new ManagedLedgerException(offloadException), ctx);
+                    }
+                });
     }
 
+    private CompletableFuture<Void> updateLedgerInfoForOffloaded(List<Pair<Long, byte[]>> contexts) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        synchronized (this) {
+            ledgersListMutex.lock();
+
+            contexts.forEach((context) -> {
+                    long ledgerId = context.getLeft();
+                    LedgerInfo oldInfo = ledgers.get(ledgerId);
+                    if (oldInfo == null) {
+                        cleanupOffloadedOnFailure(context.getLeft(), context.getRight(),
+                                                  "Trimmed while offload in progress");
+                    } else {
+                        LedgerInfo newInfo = oldInfo.toBuilder()
+                            .setOffloadContext(ByteString.copyFrom(context.getRight())).build();
+                        ledgers.put(ledgerId, newInfo);
+                    }
+                });
+
+            log.info("[{}] Updating of ledgers list after offloading", name);
+            store.asyncUpdateLedgerIds(name, getManagedLedgerInfo(), ledgersStat,
+                    new MetaStoreCallback<Void>() {
+                        @Override
+                        public void operationComplete(Void result, Stat stat) {
+                            log.info("[{}] End Offload. ledgers={} totalSize={}", name, ledgers.size(),
+                                     TOTAL_SIZE_UPDATER.get(ManagedLedgerImpl.this));
+                            ledgersStat = stat;
+                            ledgersListMutex.unlock();
+
+                            promise.complete(null);
+
+                            contexts.forEach((context) -> {
+                                    long ledgerId = context.getLeft();
+                                    ledgerCache.remove(ledgerId);
+                                    log.info("[{}] Removing offloaded ledger from bookkeeper {}",
+                                             name, ledgerId);
+                                    asyncDeleteLedger(ledgerId);
```

Comment:
I am not sure delete ledger immediately after offloading is a good idea. I would suggest separating offloading and deleting logic. offloading only handles copying data to tiered storage. the deletion logic should be happen in a different call. so that pulsar broker can actually control the amount of data stores in bk and the amount of data stores in tiered storage via its retention policy. e.g. user can configure keep X-mb or Y-ledgers offloaded ledger.

Reply:
Do you mean let retention clean up the bk ledgers? Or having a new configuration option to clean up the bk ledger some time after offloading, but before retention hits?

Reply:
I mean a new configuration to clean up bk ledger some time after offloading, but before retention hits.

Reply:
ok, ill remove all delete operations for now, and create a task for background delete.